### DISCUSSION TITLE
Add 'Read' Action with an Event to Custom API Hook

### DISCUSSION THREAD

**Discussion by jocstech** [directus/directus#3736] [Feature Requests]
**Proposal:** 

Add support to '**Read**' Action with an Event to **Custom API Hook** Extensions. 

**Use Case:** 

Implementing a trending posts feature(items sort by item's API read/clicks counts total). 

**Mock Code Example:**

```javascript
module.exports= function registerHook({ services }) {
	return {
		'items.read': function ({ collection }) {
			if (collection === 'articles') {
				// TODO: services -> update the clicks_count field +1
			}
		},
	};
};
```

or is there any other quick way to make this 'trending posts' feature available within Directus?  Thank You.

**Comment by rijkvanzanten** [comment]
> or is there any other quick way to make this 'trending posts' feature available within Directus? Thank You.

You could use a regular `response` hook and check for the GET + /items/articles endpoint

**Reply by jocstech** [reply]
Could you tell more details about 'response' hook usage, since I can not find any docs and example about it. Thanks

**Reply by rijkvanzanten** [reply]
See https://docs.directus.io/guides/extensions/api-hooks.html

**Reply by philleepflorence** [reply]
@rijkvanzanten ... can you confirm if the _read_ action is available for the _items_ scope or would one need to listen for the HTTP event within the hook?

**Reply by rijkvanzanten** [reply]
There is no read action available (yet), hence my recommendation to use the generic `response` one

**Reply by philleepflorence** [reply]
Thanks for the confirmation...

> While there is no _read_ action yet, you can register a hook function using the _response_ scope, listen for the **CRUD** HTTP Methods to perform additional custom logic.

Adding to the docs...please validate.

**Reply by MoltenCoffee** [reply]
I see this is an older discussion, but I'd like to add my two cents. The proposed solution with the `response` event won't work, as far as I know, for the GraphQL endpoint, as the endpoint is always the same.

Has there been any previous correspondence about or arguments against a `read` hook? If desired, I'd be more than happy looking into it.

**Reply by rijkvanzanten** [reply]
> Has there been any previous correspondence about or arguments against a read hook?

It's mostly a performance concern. There can be _a lot_ of reads in any given request, so based on what you do in the hook, you might easily run into performance problems

**Reply by MoltenCoffee** [reply]
True.. I'd personally argue this is the responsibility of the end user, as they are perfectly capable of obliterating performance with other hooks (especially using `before`). 
I do understand though, that `read` performance is more important than other actions for a CMS.

One option I see is emitting an action in a middleware on the respective routers instead, but I figure that could cause it's own problems (and potentially complicate the system by a lot, as the user writing the hook would have to implement logic to make sense of the request by themselves).

Interesting conundrum!

**Reply by rijkvanzanten** [reply]
> One option I see is emitting an action in a middleware on the respective routers instead, but I figure that could cause it's own problems (and potentially complicate the system by a lot, as the user writing the hook would have to implement logic to make sense of the request by themselves).

Yeah.. and that doesn't really scale across the multiple api interfaces (eg rest vs graphql vs websockets vs whatever else comes along in the future) 🤔 

I think I agree with 

> True.. I'd personally argue this is the responsibility of the end user, as they are perfectly capable of obliterating performance with other hooks (especially using before).

and be a little more hands-off in this case. If you screw up the API by adding a hook that freezes everything up, then.. well.. don't do that

**Reply by philleepflorence** [reply]
If I may suggest a data audit, analytics, and API access implementation.

- <ins>[Updated Solution](https://github.com/directus/directus/pull/6347)</ins>
- <del>A `read` action added to `directus_activity` for full **Auth-CRUD** activity.</del>
- <del>Access log in Nginx no longer logs API request information esp. `access_token` </del>
    - <del>confirmed after `directus_activity.create` event.</del>
    - <del>This also helped with a security breach related to [Mechanism to redact sensitive information in the logs]( https://github.com/directus/directus/discussions/6295)</del>

> User Story
> As an analytics admin, I want to view and analyze all events that happen within Directus tracked and stored in the activities collection to provide full accountability over everything that happens to within the API.

**Reply by MoltenCoffee** [reply]
It seems to me we're talking about several different, parallel systems. I'd personally put a `read` action in `activity` behind a feature flag in the configuration, as I think the use case is relatively limited, and can quickly clog up the logs for frequently visited (public facing) APIs. Logging read access is, as far as I know, something primarily done when working with sensitive information, like in healthcare (RFC3881, NEN7513), and may not be applicable to most users of Directus, imho.

**Reply by philleepflorence** [reply]
It is possible the use case doesn't apply to many...hence the suggestion.

> I'd personally put a `read` action in `activity` behind a feature flag in the configuration

I like and will actually borrow this idea for the few installations that don't need this!
Though I mostly build `data-driven user apps`.

**Reply by rijkvanzanten** [reply]
I wouldn't want to store every read in directus_activity, that will blow up your database real fast.

For people needing rotated access logs, I'd recommend relying on the regular stdout logs and piping them to a file using something like `pino-tee`

**Reply by philleepflorence** [reply]
> I wouldn't want to store every read in directus_activity, that will blow up your database real fast.

In D8, not a problem with Logging and [Archiving](https://www.philleepflorence.com/-/blog/extending-directus-custom-endpoint-archive-directus-activity-and-revisions-collections) rows of 550 Million a month.

That was PHP then, now with Node, your [implementation](https://github.com/directus/directus/pull/6347) provides an efficient route.

**Comment by benhaynes** [comment]
This feature has been approved, and will be tracked here: https://github.com/directus/directus/issues/10164

**Comment by rijkvanzanten** [comment]
This has since been added!
